### ISSUE TITLE
manual_map suggests wrong code due to missing autoderef

### ISSUE THREAD

**Issue by SkiFire13** [rust-lang/rust-clippy#7077]
I tried this code:

```rust
pub fn test(s: Option<&String>) -> Option<&str> {
    match s.as_ref() {
        Some(s) => Some(s),
        None => None,
    }
}
```

I got the following warning:

```
warning: manual implementation of `Option::map`
 --> src\lib.rs:2:5
  |
2 | /     match s.as_ref() {
3 | |         Some(s) => Some(s),
4 | |         None => None,
5 | |     }
  | |_____^ help: try this: `s.as_ref().map(|s| s)`
  |
  = note: `#[warn(clippy::manual_map)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_map
```

Applying the suggestion makes the code not compile anymore with the following error:

```
error[E0308]: mismatched types
 --> src\lib.rs:9:5
  |
9 |     s.as_ref().map(|s| s)
  |     ^^^^^^^^^^^^^^^^^^^^^ expected `str`, found `&std::string::String`
  |
  = note: expected enum `std::option::Option<&str>`
             found enum `std::option::Option<&&std::string::String>`
```

Took from https://github.com/rust-lang/rust/issues/83988

Happens on both stable and nightly
